### PR TITLE
TimeHelper: migrate to Java 8's `java.time` #8362 - migrate course `timeZone` field to `ZoneId`

### LINKED ISSUE #8362: TimeHelper: migrate to Java 8's `java.time`
The app is currently using the old `Date` and `Calendar` classes from `java.util` which is no longer recommended. As we have moved to Java 8, we should use the new specialized classes in `java.time` instead. It should also be easier to implement Daylight Saving Time (DST) with these new classes. From reading the source code as well as [discussions](https://github.com/TEAMMATES/teammates/issues/8094) on the issue tracker, @darrenwee and I have identified a few things that we would want to do regarding how our app handles time:

1. Refactor `TimeHelper` so that it uses the new `java.time` classes instead. Also have to modify all usage of these functions as the signature is changed.
2. Migrate the current timezone which is sometimes `double` and sometimes `String` to the `java.time`'s `ZoneId` and/or `ZoneOffset`. This would require migrating the database
3. Migrate timestamp fields in the database to UTC instead of Admin's time. This also requires migrating database. Some work on this has been done in [this PR](https://github.com/TEAMMATES/teammates/issues/8094)
4. Implement Daylight Saving Time.

For now, @darrenwee and I are working on point number 1, which will not require any modifications to the database. Here are some points that we have discussed:
- Keep using `double` for timezone in the interface, but convert it to Java's `ZoneOffset` using a private method for internal logic.
- Change the current `java.util.Date` type to `java.time.ZonedDateTime`. `java.util.Calendar` should also be converted to using `java.time.ZonedDataTime`
- For each method in `TimeHelper`, we will create a pull request to refactor that method as well as wherever the method is called.  If the data is further used by a method that we have not refactored, we will convert it back to the old `java.util.Date` temporarily until further PR refactor the method. -> Well the purpose of this is to try to keep each PR short and easier to review, but it also means we will done a lot of unnecessary conversion until everything has been migrated to Java 8's time API

@whipermr5 and @wkurniawan07 , what do you think of our approach?

### PR DESCRIPTION
**Pull request by tran-tien-dat**
Part of #8362

This PR changes the `timeZone` field in`CourseAttributes` to `ZoneId` so that other classes that need time zone information can use it directly. Since `CourseAttributes` now always contains valid time zone, validation of the user supplied time zone field is done in `CourseLogic`. This comes after the Action level so usage of `CourseAttributes` in the Action level is removed.

### CODE REVIEW

File: src/main/java/teammates/logic/api/Logic.java [hunk 1]
```diff
@@ -665,12 +665,14 @@ public CourseSummaryBundle getCourseSummaryWithFeedbackSessions(InstructorAttrib
     /**
      * Updates the details of a course.
      *
-     * @see CoursesLogic#updateCourse(CourseAttributes)
+     * @see CoursesLogic#updateCourse(String, String, String)
```
Comment (darrenwee) [inline src/main/java/teammates/logic/api/Logic.java]: What is this `@see` annotation for? I don't think it's necessary since `CoursesLogic.updateCourse(String, String, String)` is called below, so we can go into its declaration if we need to.

File: src/main/java/teammates/ui/controller/InstructorCourseAddAction.java [hunk 2]
```diff
@@ -73,8 +71,8 @@ public ActionResult execute() {
         String courseNameToShowParam = "";
 
         if (isError) { // there is error in adding the course
-            courseIdToShowParam = SanitizationHelper.sanitizeForHtml(newCourse.getId());
-            courseNameToShowParam = SanitizationHelper.sanitizeForHtml(newCourse.getName());
+            courseIdToShowParam = SanitizationHelper.sanitizeForHtml(newCourseId);
+            courseNameToShowParam = SanitizationHelper.sanitizeForHtml(newCourseName);
 
             List<String> statusMessageTexts = new ArrayList<>();
```
Comment (darrenwee) [inline src/main/java/teammates/ui/controller/InstructorCourseAddAction.java]: It's not touched by your PR, but can you refactor this `List` and `for` part to use a stream since we're moving to java 8 anyway?

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 3]
```diff
@@ -718,4 +718,24 @@ public boolean hasIndicatedSections(String courseId) throws EntityDoesNotExistEx
...
+
+        FieldValidator validator = new FieldValidator();
+        String timeZoneErrorMessage = validator.getInvalidityInfoForCourseTimeZone(courseTimeZone);
+        if (!timeZoneErrorMessage.isEmpty()) {
+            CourseAttributes dummyCourse = CourseAttributes
+                    .builder(courseId, courseName, ZoneId.of("UTC"))
+                    .build();
+            List<String> errors = dummyCourse.getInvalidityInfo();
+            errors.add(timeZoneErrorMessage);
+            throw new InvalidParametersException(errors);
```
Comment (darrenwee) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: I am not clear on what this if block is doing. This seems like a very roundabout way of handing an invalid timezone. Why is `getInvalidityInfo` only called when there is a `timeZoneErrorMessage`? What if there are other issues with the parameters, but the `courseTimeZone` field is correct?

File: src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java [hunk 4]
```diff
@@ -41,7 +41,7 @@
      * <li>{@code createdAt = current date}</li>
      * </ul>
      */
-    public static Builder builder(String courseId, String name, String timeZone) {
+    public static Builder builder(String courseId, String name, ZoneId timeZone) {
```
Comment (darrenwee) [inline src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java]: Can you rewrite the javadoc for this method to follow the coding standard?

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 5]
```diff
@@ -718,4 +718,24 @@ public boolean hasIndicatedSections(String courseId) throws EntityDoesNotExistEx
...
+
+    private CourseAttributes validateAndCreateCourseAttributes(String courseId,
+                                                               String courseName, String courseTimeZone)
+            throws InvalidParametersException {
+
+        FieldValidator validator = new FieldValidator();
+        String timeZoneErrorMessage = validator.getInvalidityInfoForCourseTimeZone(courseTimeZone);
+        if (!timeZoneErrorMessage.isEmpty()) {
+            CourseAttributes dummyCourse = CourseAttributes
+                    .builder(courseId, courseName, ZoneId.of("UTC"))
```
Comment (darrenwee) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Can use `Const.DEFAULT_TIMEZONE` to replace `"UTC"`?

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 6]
```diff
@@ -111,7 +112,7 @@ public void deleteCourse(String courseId) {
 
         // only the courseId is important here, everything else are placeholders
         deleteEntity(CourseAttributes
-                .builder(courseId, "Non-existent course", "UTC")
+                .builder(courseId, "Non-existent course", ZoneId.of("UTC"))
```
Comment (darrenwee) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Same as above about `Const.DEFAULT_TIMEZONE`.

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 7]
```diff
@@ -126,7 +128,7 @@ private void testContent() throws Exception {
         homePage.verifyHtmlMainContent("/instructorHomeNewInstructorWithoutSampleCourse.html");
 
         CourseAttributes newCourse = CourseAttributes
-                .builder("newIns.wit-demo", "Sample Course 101", "UTC")
+                .builder("newIns.wit-demo", "Sample Course 101", ZoneId.of("UTC"))
```
Comment (darrenwee) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Same as above for `Const.DEFAULT_TIMEZONE`.

File: src/test/java/teammates/test/cases/datatransfer/CourseAttributesTest.java [hunk 8]
```diff
@@ -17,7 +18,7 @@
 
     private String validName = "validName";
     private String validId = "validId";
-    private String validTimeZone = "validTimeZone";
+    private ZoneId validTimeZone = ZoneId.of("UTC");
```
Comment (darrenwee) [inline src/test/java/teammates/test/cases/datatransfer/CourseAttributesTest.java]: Same as above for `Const.DEFAULT_TIMEZONE`.

File: src/main/java/teammates/logic/api/Logic.java (hunk 1 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/api/Logic.java]: Just following the old javadoc. I think it's meant for when the javadoc is generated as html?

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 5 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: I think we should remove `Const.DEFAULT_TIMEZONE`. Times are recorded as instants which has no associated timezone, so i don't see how we should have this constant.

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 3 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: Yeah i should have explained this part more. If you follow the flow, parameters checking will be done in `CourseDb` by calling `CourseAtrributes.getInvalidityInfo`. However it can only check for id and name, because once a `CourseAtrributes` is created, its time zone is automatically valid. Hence I have to check for the timezone beforehand here. If it fails, i mimick the checking of the other fields to add to the list of errors. Otherwise I leave the checking of the other fields to `CourseDb`, as how it's originally done.

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java (hunk 7 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: This is a test. It should explicitly states what timezone it's using

File: src/test/java/teammates/test/cases/datatransfer/CourseAttributesTest.java (hunk 8 above)
Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/datatransfer/CourseAttributesTest.java]: This is a test. It should explicitly states what timezone it's using

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 5 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: Anyway it's just a valid dummy timezone so I don't see why we need a constant for it.

File: src/main/java/teammates/ui/controller/InstructorCourseAddAction.java (hunk 2 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/ui/controller/InstructorCourseAddAction.java]: I think it should be left to other PRs because this appears in many Action classes and it should be standardized, i.e. Either leave them the way they are or change all of them at once.

File: src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java (hunk 4 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java]: Sorry what's the coding standard you are referring to?

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 5 above)
Comment (darrenwee) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: Ah, my bad. I thought you wanted to use it because you fixed it to `"UTC"` in #8540 that just got merged.

File: src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java (hunk 4 above)
Comment (darrenwee) [reply src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java]: https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html#comments (from dev docs for teammates)

Comment (tran-tien-dat) [reply src/main/java/teammates/common/datatransfer/attributes/CourseAttributes.java]: Thanks. So just add descriptions of the parameters?

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 9]
```diff
@@ -718,4 +718,24 @@ public boolean hasIndicatedSections(String courseId) throws EntityDoesNotExistEx
         }
         return archivedCourseIds;
     }
+
+    private CourseAttributes validateAndCreateCourseAttributes(String courseId,
+                                                               String courseName, String courseTimeZone)
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Use 8-space indentation?

File: src/main/java/teammates/ui/controller/InstructorCourseAddAction.java [hunk 10]
```diff
@@ -109,10 +106,6 @@ private void createCourse(CourseAttributes course) {
         } catch (InvalidParametersException e) {
             setStatusForException(e);
         }
-
-        if (isError) {
-            return;
-        }
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseAddAction.java]: Unrelated change? Seems like `isError` is never set to true anywhere, but was never set to true even in the previous code, so this is unrelated.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 11]
```diff
@@ -68,9 +69,7 @@ public static CoursesLogic inst() {
     public void createCourse(String courseId, String courseName, String courseTimeZone)
             throws InvalidParametersException, EntityAlreadyExistsException {
 
-        CourseAttributes courseToAdd = CourseAttributes
-                .builder(courseId, courseName, courseTimeZone)
-                .build();
+        CourseAttributes courseToAdd = validateAndCreateCourseAttributes(courseId, courseName, courseTimeZone);
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Update `testCreateCourse` in `CoursesLogicTest` to test invalid time zone behaviour, now that this is not covered by `CourseAttributesTest`.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 12]
```diff
@@ -595,12 +594,13 @@ public CourseSummaryBundle getCourseSummaryWithoutStats(String courseId) throws
...
+     * @param courseName new name of the course
+     * @param courseTimeZone new time zone of the course
      */
-    public void updateCourse(CourseAttributes newCourse) throws InvalidParametersException,
-                                                                EntityDoesNotExistException {
-        Assumption.assertNotNull(Const.StatusCodes.NULL_PARAMETER, newCourse);
-
+    public void updateCourse(String courseId, String courseName, String courseTimeZone)
+            throws InvalidParametersException, EntityDoesNotExistException {
+        CourseAttributes newCourse = validateAndCreateCourseAttributes(courseId, courseName, courseTimeZone);
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Write a test `testUpdateCourse` in `CoursesLogicTest` to test invalid time zone behaviour, now that this is not covered by `CourseAttributesTest`.

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 3 above)
Comment (whipermr5) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: Might want to add some brief explanation in code comments to avoid future confusion?

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 11 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: Ok

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 12 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: ok

File: src/main/java/teammates/common/util/JsonUtils.java [hunk 13]
```diff
@@ -128,4 +130,21 @@ public synchronized Instant deserialize(JsonElement element, Type type, JsonDese
...
+        @Override
+        public synchronized JsonElement serialize(ZoneId zoneId, Type type, JsonSerializationContext context) {
+            return new JsonPrimitive(zoneId.getId());
+        }
+
+        @Override
+        public synchronized ZoneId deserialize(JsonElement element, Type type, JsonDeserializationContext context) {
+            try {
+                return ZoneId.of(element.getAsString());
+            } catch (DateTimeParseException e) {
```
[the remark this answers is not in the document]
Comment (tran-tien-dat) [reply src/main/java/teammates/common/util/JsonUtils.java]: Yep my bad.

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 9 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: ok

File: src/main/java/teammates/logic/core/CoursesLogic.java (hunk 3 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: Yep, will do.

File: src/main/java/teammates/ui/controller/InstructorCourseAddAction.java (hunk 10 above)
Comment (tran-tien-dat) [reply src/main/java/teammates/ui/controller/InstructorCourseAddAction.java]: It's set to true in `setStatusForException`. But anyway, it's unrelated to why I removed this part. I remove it because it does nothing. Whether `isError` is true or false the function also just return

Comment (whipermr5) [reply src/main/java/teammates/ui/controller/InstructorCourseAddAction.java]: Ohh that's true; didn't realise 😅 It's fine then.

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 14]
```diff
@@ -148,7 +149,7 @@ private Course getCourseEntity(String courseId) {
     protected CourseAttributes makeAttributes(Course entity) {
         Assumption.assertNotNull(Const.StatusCodes.DBLEVEL_NULL_INPUT, entity);
 
-        return CourseAttributes.builder(entity.getUniqueId(), entity.getName(), entity.getTimeZone())
+        return CourseAttributes.builder(entity.getUniqueId(), entity.getName(), ZoneId.of(entity.getTimeZone()))
```
[the remark this answers is not in the document]
Comment (whipermr5) [reply src/main/java/teammates/storage/api/CoursesDb.java]: Actually, maybe do it in `Course` using `@OnLoad`, since it's only needed temporarily. After data migration it can be removed.

**Comment by whipermr5** [comment]
Btw, the assignees to a PR should be the reviewer(s), not the author.

File: src/main/java/teammates/storage/api/CoursesDb.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tran-tien-dat) [reply src/main/java/teammates/storage/api/CoursesDb.java]: Data is already using valid timezone string, so there's no need for migration? Anyway, I think it should be done here no matter what, in case the app fails somewhere and invalid timezone strings appear in the database.

[the remark this answers is not in the document]
Comment (whipermr5) [reply src/main/java/teammates/storage/api/CoursesDb.java]: I was referring to `GMT+0`, `GMT-0` etc. that were supported by Joda time but not java.time. But you have a point - doing it here is good for fault tolerance. Let's do it here then.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 15]
```diff
@@ -719,21 +719,33 @@ public boolean hasIndicatedSections(String courseId) throws EntityDoesNotExistEx
         return archivedCourseIds;
     }
 
-    private CourseAttributes validateAndCreateCourseAttributes(String courseId,
-                                                               String courseName, String courseTimeZone)
-            throws InvalidParametersException {
+    /**
+     * Check that {@code courseTimeZone} is valid and then return a {@code CourseAttributes}
+     *
+     * Field validation is usually done in {@code CourseDb} by calling {@code CourseAttributes.getInvalidityInfo}.
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Use `@link` for all references to classes/methods.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 16]
```diff
@@ -719,21 +719,33 @@ public boolean hasIndicatedSections(String courseId) throws EntityDoesNotExistEx
...
             CourseAttributes dummyCourse = CourseAttributes
                     .builder(courseId, courseName, ZoneId.of("UTC"))
                     .build();
             List<String> errors = dummyCourse.getInvalidityInfo();
             errors.add(timeZoneErrorMessage);
+            // Imitate exception throwing in `CourseDb`
             throw new InvalidParametersException(errors);
         }
 
+        // If time zone field is valid, leave validation  of other fields to `CourseDb` like usual
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: `CoursesDb`

File: src/test/java/teammates/test/cases/logic/CoursesLogicTest.java [hunk 17]
```diff
@@ -979,4 +989,24 @@ private void testDeleteCourse() {
             assertEquals(Const.StatusCodes.DBLEVEL_NULL_INPUT, e.getMessage());
         }
     }
+
+    private void testUpdateCourse() throws Exception {
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/logic/CoursesLogicTest.java]: Maybe write a typical case too?

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 18]
```diff
@@ -718,4 +718,35 @@ public boolean hasIndicatedSections(String courseId) throws EntityDoesNotExistEx
         }
         return archivedCourseIds;
     }
+
+    /**
+     * Check that {@code courseTimeZone} is valid and then return a {@link CourseAttributes}
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Check**s**, return**s**. End the sentence with a full stop.

Comment (tran-tien-dat) [reply src/main/java/teammates/logic/core/CoursesLogic.java]: My apology for not following the javadoc standard. Will take note next time.

File: src/test/java/teammates/test/cases/logic/CoursesLogicTest.java [hunk 19]
```diff
@@ -775,26 +778,34 @@ private void testHasIndicatedSections() throws Exception {
...
         try {
-            coursesLogic.createCourse(null, c.getName(), c.getTimeZone());
+            coursesLogic.createCourse(null, c.getName(), c.getTimeZone().getId());
             signalFailureToDetectException();
         } catch (AssertionError e) {
             assertEquals("Non-null value expected", e.getMessage());
         }
+        ______TS("Invalid time zone");
+
+        String invalidTimeZone = "Invalid Timezone";
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/logic/CoursesLogicTest.java]: Test invalid ID too together with invalid time zone to check that both error messages are included.

File: src/main/java/teammates/storage/api/CoursesDb.java [hunk 20]
```diff
@@ -148,7 +150,15 @@ private Course getCourseEntity(String courseId) {
     protected CourseAttributes makeAttributes(Course entity) {
         Assumption.assertNotNull(Const.StatusCodes.DBLEVEL_NULL_INPUT, entity);
 
-        return CourseAttributes.builder(entity.getUniqueId(), entity.getName(), entity.getTimeZone())
+        ZoneId courseTimeZone;
+        try {
+            courseTimeZone = ZoneId.of(entity.getTimeZone());
+        } catch (DateTimeException e) {
+            log.severe("Timezone '" + entity.getTimeZone() + "' of course '" + entity.getUniqueId()
+                    + "' is no longer supported. UTC will be used instead.");
```
Comment (whipermr5) [inline src/main/java/teammates/storage/api/CoursesDb.java]: Change `no longer` to `not` so that the message is relevant for completely invalid data too.

File: src/test/java/teammates/test/cases/logic/CoursesLogicTest.java [hunk 21]
```diff
@@ -28,6 +28,7 @@
 import teammates.storage.api.CoursesDb;
 import teammates.storage.api.InstructorsDb;
 import teammates.test.driver.AssertHelper;
+import teammates.test.driver.StringHelperExtension;
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/logic/CoursesLogicTest.java]: I think this will cause the build to fail

Comment (tran-tien-dat) [reply src/test/java/teammates/test/cases/logic/CoursesLogicTest.java]: My bad, it comes when i copied the code from `CourseAttributesTest` but I forgot to remove.
